### PR TITLE
Refactor parameter validation into method with explanatory comment

### PR DESCRIPTION
**Pull request by michaeldiamant**
Refactors parameter type checking validation in #198 into a separate method with explanatory note about implicit `Expr` declaration.

Consider the PR optional.  If we don't feel it's an improvement, feel welcomed to close it.

Motivation:  
* During PR review, I opened a question on the topic:  https://github.com/algorand/pyteal/pull/198#discussion_r815906297.  
* Since the implicit declaration topic took me some time to understand, I supplied an inline comment.
* The method refactoring is optional.  I provided it to eliminate mutation of `expected_arg_type` and secondarily, to shorten constructor definition.

### CODE REVIEW

**Comment by tzaffi** [comment]
This is a clear improvement, but let's get tests to pass.

File: pyteal/ast/subroutine.py [hunk 1]
```diff
@@ -95,6 +76,36 @@ def __init__(
         self.declaration: Optional["SubroutineDeclaration"] = None
         self.__name = self.implementation.__name__ if nameStr is None else nameStr
 
+    @staticmethod
+    def _validate_parameter_type(user_defined_annotations: OrderedDict, parameter_name: str):
```
Comment (tzaffi) [inline pyteal/ast/subroutine.py:80]:
```suggestion
    def _validate_parameter_type(user_defined_annotations: OrderedDict, parameter_name: str) -> type:
```

**Review by tzaffi — COMMENTED** [review]
asking for return type annotation as well

File: pyteal/ast/subroutine.py [hunk 2]
```diff
@@ -95,6 +76,38 @@ def __init__(
         self.declaration: Optional["SubroutineDeclaration"] = None
         self.__name = self.implementation.__name__ if nameStr is None else nameStr
 
+    @staticmethod
+    def _validate_parameter_type(
+        user_defined_annotations: dict[str, str], parameter_name: str
+    ):
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply pyteal/ast/subroutine.py:82]: may need to add import up top:
```python
from typing import Dict
```

[the remark this answers is not in the document]
Comment (tzaffi) [reply pyteal/ast/subroutine.py]: hmmm... worth a try to see if you can just use the `OrderedDict` from collections and not import from typing... Haven't tried that.

[the remark this answers is not in the document]
Comment (michaeldiamant) [reply pyteal/ast/subroutine.py]: @tzaffi Adding a return type annotation makes sense.  Sorry, I missed it.  I added it via https://github.com/algorand/pyteal/pull/217/commits/59c2dc424bdd0357995901736cc73f7fe5a415a1.

File: pyteal/ast/subroutine.py [hunk 3]
```diff
@@ -95,6 +76,38 @@ def __init__(
         self.declaration: Optional["SubroutineDeclaration"] = None
         self.__name = self.implementation.__name__ if nameStr is None else nameStr
 
+    @staticmethod
+    def _validate_parameter_type(
+        user_defined_annotations: dict, parameter_name: str
+    ) -> Type[Expr]:
```
Comment (tzaffi) [inline pyteal/ast/subroutine.py:82]:
```suggestion
    ) -> Type[Union[Expr, ScratchVar]]:
```

**Review by tzaffi — CHANGES_REQUESTED** [review]
request change to conform with line 32 of `subroutine.py`

File: pyteal/ast/subroutine.py (hunk 3 above)
Comment (michaeldiamant) [reply pyteal/ast/subroutine.py]: @tzaffi Thanks for the assist!  ☕

**Review by tzaffi — APPROVED** [review]
LGTM

**Pull request merged**